README.md
# Bit-Twist: Libpcap-based Ethernet packet generator

SPDX-License-Identifier: GPL-2.0-or-later

Supported systems: Linux, BSD, macOS, and Windows.

Bit-Twist is a simple yet powerful libpcap-based Ethernet packet generator and
packet editor. It is designed to complement tcpdump, which by itself has done a
great job at capturing network traffic.

> With Bit-Twist, you can now regenerate your captured traffic onto a live
> network! Packets are generated from tcpdump trace file (.pcap file).
> Bit-Twist also comes with a comprehensive trace file editor to allow you to
> change the contents of a trace file.

> Bit-Twist is designed for exceptional speed! Utilizing a standard laptop with
> Linux system, you can edit **10 million packets in less than 5 seconds**, and
> send the packets onto a live network, achieving throughput levels that match
> the maximum line rate of your NIC.

Packet generator is useful in simulating networking traffic or scenario,
testing firewall, IDS, and IPS, and troubleshooting various network problems.

## Features

These are just a few significant features that makes Bit-Twist unique and
stands out as one of the best Ethernet packet generator and packet editor
package made available to the open source community.

- Highly portable: Bit-Twist runs on Linux, BSD, macOS, and Windows.

- Send multiple trace files indefinitely with set interval, packets per second,
  or line rate between 1 Mbps to 10 Gbps using built-in token bucket algorithm.

- Comprehensive trace file editor to edit most fields in Ethernet, ARP, IPv4,
  IPv6, ICMPv4, ICMPv6, TCP, and UDP headers. Templates are also included to
  generate packets with these headers without needing an existing trace file.

- Automatic header checksum correction (with option to disable).

- Send packets with custom QoS bits to test classification and queuing features
  of switches and routers.

- Send packets with uniformly distributed random numbers for port numbers,
  TCP sequence numbers, etc.

- Append custom payload (e.g. copy of hex stream from Wireshark) to existing
  packets after a specific header, handy for testing new protocols.

- Send packets with truncated or expanded length in bytes, facilitating
  incremental throughput testing.

- Highly scriptable: With proper manipulation, you can turn Bit-Twist into a
  versatile packet generator and packet editor tooling to meet your network
  testing requirements.

For the complete feature list, see Bit-Twist man pages:

- [bittwist.1](https://bittwist.sourceforge.io/doc/bittwist.1.html) - pcap based ethernet packet generator
- [bittwiste.1](https://bittwist.sourceforge.io/doc/bittwiste.1.html) - pcap capture file editor

## Examples

Please visit https://bittwist.sourceforge.io/doc.html for examples on how to
use Bit-Twist.

## Installation

Follow the instructions for your operating system below to install, run, or
build Bit-Twist on your machine.

Unless specified otherwise:

- executables (bittwist, bittwiste) are installed in /usr/local/bin
- manual pages (bittwist.1, bittwiste.1) are installed in /usr/local/share/man/man1

For more general information, please visit https://bittwist.sourceforge.io

## For Windows systems

This distribution is tested to work on Microsoft Windows 10.

### Installation

- Download Npcap installer from https://npcap.com/dist/npcap-1.75.exe

- Run the Npcap installer to install Npcap on your system.
  Select "Install Npcap in WinPcap API-compatible Mode" option during the
  installation.

- Extract bittwist-windows-3.8.zip into C:\Users\_YOUR_USERNAME_\Downloads

- In Command Prompt:

```
> cd C:\Users\_YOUR_USERNAME_\Downloads\bittwist-windows-3.8\src
> bittwist -h   (usage for packet generator)
> bittwiste -h  (usage for packet editor)
> bittwist -d   (to view available network cards you can send packets on)
```

- You may readily use Bit-Twist from the src directory as above.

- If you wish to install Bit-Twist system-wide, copy the files from src
  directory into C:\WINDOWS\system32:
  - bittwist.exe
  - bittwist.exe
  - cygwin1.dll (From https://www.cygwin.com)

- Manual pages are available in doc/

### Recompilation

This distribution is compiled against Npcap 1.75 with Npcap SDK 1.13 in
Cygwin environment on Microsoft Windows 10.

If you wish to rebuild Bit-Twist from source files, you will need Cygwin
environment:

- Download Cygwin installer from https://www.cygwin.com/setup-x86_64.exe

- Run the Cygwin installer to install Cygwin environment on your system.

- Be sure to select at least the following packages in the
  "Cygwin Setup - Select Packages" window:
  - Devel > gcc-core
  - Devel > make
  - Devel > binutils 2.38-1 (newer version may not work yet)
  - Devel > mingw64-x86_64-binutils 2.38-1 (newer version may not work yet)

- Click on the "Cygwin64 Terminal" icon in your desktop to launch a new
  terminal under Cygwin environment.

```
$ cd /cygdrive/c/Users/_YOUR_USERNAME_/Downloads/bittwist-windows-3.8
$ make
$ make install
$ bittwist -h   (usage for packet generator)
$ bittwiste -h  (usage for packet editor)
$ bittwist -d   (to view available network cards you can send packets on)
```

## For Linux systems

This distribution is tested to work on CentOS Stream 9.

### Required dependencies

- Libpcap is required (available for download from https://www.tcpdump.org/).
  This distribution is compiled against libpcap 1.10.4.
  Sample installation of libpcap 1.10.4 on CentOS Stream 9:

```
$ sudo yum install make gcc flex bison
$ wget https://www.tcpdump.org/release/libpcap-1.10.4.tar.gz
$ tar -xzf libpcap-1.10.4.tar.gz
$ cd libpcap-1.10.4
$ ./configure && make && sudo make install
$ sudo ldconfig
```

### Installation

```
$ tar -xzf bittwist-linux-3.8.tar.gz
$ cd bittwist-linux-3.8
$ make
$ sudo make install
```

## For macOS systems

This distribution is tested to work on macOS Ventura 13.3.1

### Required dependencies

- Libpcap is required (available for download from https://www.tcpdump.org/).
  This distribution is compiled against libpcap 1.10.4. However, any existing
  libpcap on your system may work normally with Bit-Twist.

- Xcode command line developer tools; you will be prompted to install this
  automatically on your first attempt to run make below.

### Installation

```
$ tar -xzf bittwist-macos-3.8.tar.gz
$ cd bittwist-macos-3.8
$ make
$ sudo make install
```

## For BSD systems

This distribution is tested to work on FreeBSD 13.2-RELEASE.

### Required dependencies

- Libpcap is required (available for download from https://www.tcpdump.org/).
  This distribution is compiled against libpcap 1.10.4. However, any existing
  libpcap on your system may work normally with Bit-Twist.

### Installation

```
$ tar -xzf bittwist-bsd-3.8.tar.gz
$ cd bittwist-bsd-3.8
$ make
$ sudo make install
```


CHANGES
06/07/2023: Version 3.8
- bumped from gcc -O2 to -O3 optimization flag:
  https://gcc.gnu.org/onlinedocs/gcc/Optimize-Options.html#index-O3
bittwist (generator)
- added RFC 2544 benchmark data generator and benchmark page
- improved actual inter-packet gap replay by using token bucket algorithm for
  short IPG and self-regulated timer for long IPG

27/06/2023: Version 3.7
bittwist (generator)
- 30% speed up when sending large pcap file
bittwiste (editor)
- added -G option to set fixed or random inter-packet gap in microseconds
- added support for CIDR notation when modifying IPv4/IPv6 address
- added support for random MAC address

17/06/2023: Version 3.6
- added github repo for source code browsing:
  https://github.com/ayeowch/bittwist
bittwist (generator)
- fixed linerate throttling
- updated -p option to send packets at set packets per second (pps)
- added -t option to set inter-packet gap (this was previously -p option)
bittwiste (editor)
- added option to edit IPv6 next header
- added random number support for id, ttl, proto, next header, hop limit

16/06/2023: Version 3.5
bittwiste (editor)
- added option to set uniformly distributed random number (mersenne twister)
  for tcp/udp source/destination port and tcp sequence/acknowledgment number
- added -N flag to duplicate packets from trace file, e.g.
  -N 100000 to repeat packets from trace file for 100,000 times
- added ip6tcp and ip6udp templates

13/06/2023: Version 3.4
bittwiste (editor)
- updated -I flag to allow loading trace file from built-in templates

12/06/2023: Version 3.3
bittwiste (editor)
- options to edit IPv4/IPv6 packets to allow testing of classification/queuing
  features of switches/routers, e.g.
  -T ip -e 3 to indicate congestion to the end hosts
  -T ip6 -c 16 to classify packet for operation and management of the network
  -T ip6 -f 0xfffff to set custom flow label

10/06/2023: Version 3.2
- added support for trace files with nanosecond resolution
- updated manual pages
bittwist (generator)
- removed speed (-m) flag; use -p/-r to shape throughput
- updated linerate (-r) flag to use token bucket algorithm
bittwiste (editor)
- added support for IPv6 packets
- improved cksum()
- expanded test cases

30/05/2023: Version 3.1
- use in-memory trace files when sending packets in loop
- fixed integer overflow in stats output
- added static code analysis in Makefile (make check)
- reformatted source files for C17 standard
- added initial python testing framework for Linux systems

24/05/2023: Version 3.0
- Linux distribution: rebuilt with libpcap 1.10.4
- BSD distribution: rebuilt with libpcap 1.10.4
- macOS distribution: rebuilt with libpcap 1.10.4
- Windows distribution: rebuilt with Npcap 1.75 with Npcap SDK 1.13
- updated homepage url: https://bittwist.sourceforge.io

21/04/2012: Version 2.0
- switched to "struct pcap_sf_pkthdr" to support 64-bit platforms
- rebuilt with libpcap 1.2.1
- added Mac OS X distribution (author's primary platform)
- BSD distribution no longer maintained
- ended Bit-Twist Web Edition BETA
- revised copyright information and manpages
bittwistb (bridge)
- no longer maintained on all platforms

12/12/2009: Version 1.1
- rebuilt with libpcap 1.0.0
- revised copyright information and manpages
- removed Eclipse dependency; requires only Cygwin environment for Windows
  development
- release of bittwist(w), Bit-Twist Web Edition BETA,
  http://twitter.com/bittwistw

26/10/2007: Version 1.0
- revised AUTHORS, INSTALL, copyright information and manpages
- unified source files for all supported platforms
- moved Windows development to Eclipse IDE for C/C++ Developers + Cygwin
  environment
bittwist (generator)
- increased max. packet throughput from 1Gbps to 10Gbps (option -r)
bittwiste (editor)
- corrected usage of strptime()
- use comma (,) instead of slash (/) as delimiter for all replace options
- added a feature request;
  replace a specific MAC address with a new MAC address for eth, arp
- added a feature request;
  replace a specific IP address with a new IP address for arp

23/10/2006: Version 0.80
bittwist (generator)
- improved performance;
  use single buffer for all packet data,
  calculate linerate_interval only when necessary
bittwiste (editor)
- corrected several malloc related bugs
- added a feature request;
  replace the linktype stored in the pcap file header,
  e.g. to replace raw IP linktype to Ethernet linktype,
  bittwiste -I rawip.pcap -O rawip.1.pcap -M 1
- added a feature request;
  support for writing packets within a specific timeframe only,
  e.g. bittwist -I trace.pcap -O trace.1.pcap -S
  22/10/2006,21:47:35-24/10/2006,13:16:05

17/09/2006: Version 0.74
bittwiste (editor)
- minor bug correction
- added a feature request;
  delete a specific byte offset from a packet,
  bittwiste -I trace.pcap -O trace.1.pcap -D 1-14
- added a feature request;
  replace a specific IP address with a new IP address,
  e.g. to replace source IP address 192.168.0.1 with 192.168.1.1,
  bittwiste -I ip.pcap -O ip.1.pcap -T ip -s 192.168.0.1:192.168.1.1

28/07/2006: Version 0.73
- minor correction in Makefile
- merged 180506-0.72-def.h-patch-1 (patch for OpenBSD) into BSD distribution
bittwiste (editor)
- minor bug correction
- added a feature request;
  replace a specific port number with a new port number,
  e.g. to replace destination port 1066 with 1067,
  bittwiste -I tcp.pcap -O tcp.1.pcap -T tcp -d 1066:1067
misc.
- Bit-Twist 0.72 for Windows has been released prior to this update

29/04/2006: Version 0.72
- removed definition of in_addr_t in def.h for BSD distribution
bittwistb (bridge)
- minor bug correction

28/04/2006: Version 0.71
bittwist (generator)
- removed maths.h which was previously required for round(),
  ROUND() is now defined in def.h
- minor bug correction
bittwistb (bridge)

src/def.h
/*
 * SPDX-License-Identifier: GPL-2.0-or-later
 *
 * def.h - Definition header file for Bit-Twist project
 * Copyright (C) 2006 - 2023 Addy Yeow <ayeowch@gmail.com>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

#ifndef _DEF_H_
#define _DEF_H_

#include <ctype.h>
#include <errno.h>
#include <ifaddrs.h>
#include <limits.h>
#include <poll.h>
#include <signal.h>
#include <stdarg.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#define __USE_XOPEN /* using strptime from time.h */
#include <time.h>
#include <unistd.h>
#define _NET_IF_ARP_H_ /* OpenBSD's if.h takes in if_arp.h */
#include <arpa/inet.h>
#include <net/if.h>
#include <sys/ioctl.h>
#include <sys/socket.h>
#ifdef __BSD_VISIBLE /* Linux does not have net/if_dl.h */
#include <net/if_dl.h>
#endif
#include <pcap.h>

struct pcap_timeval
{
    bpf_int32 tv_sec; /* seconds (WARNING: will overflow by 2038-01-19 03:14:07 UTC) */
    /*
     * PCAP_MAGIC: 6-digit microseconds
     * NSEC_PCAP_MAGIC: 9-digit nanoseconds
     */
    bpf_int32 tv_usec;
};

struct pcap_sf_pkthdr
{
    struct pcap_timeval ts; /* timestamp */
    bpf_u_int32 caplen;     /* length of portion present */
    bpf_u_int32 len;        /* length this packet (off wire) */
};

#define BITTWIST_VERSION "3.8"
#define BITTWISTE_VERSION BITTWIST_VERSION

#define ETH_ADDR_LEN 6   /* Ethernet address length */
#define ETH_HDR_LEN 14   /* Ethernet header length */
#define ETH_MAX_LEN 1514 /* maximum frame length, excluding CRC */
#define ARP_HDR_LEN 28   /* Ethernet ARP header length */
#define IP_ADDR_LEN 4    /* IP address length */
#define IP_HDR_LEN 20    /* default IP header length */
#define IP6_HDR_LEN 40   /* default IPv6 header length */
#define ICMP_HDR_LEN 4   /* ICMP header length (up to checksum field only) */
#define ICMP6_HDR_LEN 4  /* ICMPv6 header length (up to checksum field only) */
#define TCP_HDR_LEN 20   /* default TCP header length */
#define UDP_HDR_LEN 8    /* UDP header length */

#define ETH_TYPE_IP 0x0800   /* IP protocol */
#define ETH_TYPE_IPV6 0x86dd /* IPv6 protocol */
#define ETH_TYPE_ARP 0x0806  /* address resolution protocol */

#ifndef IPPROTO_ICMP
#define IPPROTO_ICMP 1 /* internet control message protocol */
#endif

#ifndef IPPROTO_TCP
#define IPPROTO_TCP 6 /* transmission control protocol */
#endif

#ifndef IPPROTO_UDP
#define IPPROTO_UDP 17 /* user datagram protocol */
#endif

#ifndef IPPROTO_ICMPV6
#define IPPROTO_ICMPV6 58 /* internet control message protocol version 6 */
#endif

/* bittwist */
#define LINERATE_MIN 0     /* Mbps (0 means no limit) */
#define LINERATE_MAX 10000 /* Mbps */
#define PKT_PAD 0x00       /* packet padding */
#define PPS_MAX 1000000    /* max. packets per second */
#define GAP_MAX 86400      /* arbitrary maximum inter-packet gap in seconds */

/* bittwiste */
#define FIELD_SET 1          /* flag to overwrite field with new value */
#define FIELD_REPLACE 2      /* flag to overwrite matching value in field with new value */
#define FIELD_SET_RAND 3     /* flag to overwrite field with random value */
#define FIELD_REPLACE_RAND 4 /* flag to overwrite matching value in field with random value */
#define PAYLOAD_MAX 1500     /* maximum payload in bytes */
#define GAP_START 946684800  /* arbitrary start time when using custom inter-packet gap */

/* supported header specification (dummy values) */
#define ETH 1
#define ARP 2
#define IP 3
#define IP6 30
#define ICMP 4
#define ICMP6 40
#define TCP 5
#define UDP 6

#define IP_FO_MAX 7770             /* maximum IP fragment offset (number of 64-bit segments) */
#define IP6_FLOW_LABEL_MAX 1048575 /* 20-bit flow label: 0x00000 to 0xfffff (1048575) */
#define DS_FIELD_MAX 63            /* 6-bit DS field */
#define ECN_FIELD_MAX 3            /* 2-bit ECN field */

#define PCAP_PREAMBLE_LEN 24       /* pcap file header length */
#define PCAP_HDR_LEN 16            /* pcap header length */
#define PCAP_MAGIC 0xa1b2c3d4      /* pcap magic number (timestamps in microsecond resolution)*/
#define NSEC_PCAP_MAGIC 0xa1b23c4d /* pcap magic number (timestamps in nanosecond resolution) */

#ifndef timespecisset
#define timespecisset(tsp) ((tsp)->tv_sec || (tsp)->tv_nsec)
#endif

#ifndef timespeccmp
#define timespeccmp(ctsp, ptsp, cmp)                                                               \
    (((ctsp)->tv_sec == (ptsp)->tv_sec) ? ((ctsp)->tv_nsec cmp(ptsp)->tv_nsec)                     \
                                        : ((ctsp)->tv_sec cmp(ptsp)->tv_sec))
#endif

#ifndef timespecsub
#define timespecsub(ctsp, ptsp, vtsp)                                                              \
    do                                                                                             \
    {                                                                                              \
        (vtsp)->tv_sec = (ctsp)->tv_sec - (ptsp)->tv_sec;                                          \
        (vtsp)->tv_nsec = (ctsp)->tv_nsec - (ptsp)->tv_nsec;                                       \
        if ((vtsp)->tv_nsec < 0)                                                                   \
        {                                                                                          \
            (vtsp)->tv_sec--;                                                                      \
            (vtsp)->tv_nsec += 1000000000L;                                                        \
        }                                                                                          \
    } while (0)
#endif

#define pcap_timeval_usadd(ts, us)                                                                 \
    do                                                                                             \
    {                                                                                              \
        uint64_t total_usec = (ts)->tv_usec + (us);                                                \
        (ts)->tv_sec += total_usec / 1000000;                                                      \
        (ts)->tv_usec = total_usec % 1000000;                                                      \
    } while (0)

#define pcap_timeval_nsadd(ts, ns)                                                                 \
    do                                                                                             \
    {                                                                                              \
        uint64_t total_usec = (ts)->tv_usec + (ns);                                                \
        (ts)->tv_sec += total_usec / 1000000000L;                                                  \
        (ts)->tv_usec = total_usec % 1000000000L;                                                  \
    } while (0)

/* Ethernet header */
struct ethhdr
{
    uint8_t eth_dhost[ETH_ADDR_LEN];
    uint8_t eth_shost[ETH_ADDR_LEN];
    uint16_t eth_type;
};

/* Ethernet ARP header */
struct arphdr
{
    uint16_t ar_hrd;              /* format of hardware address */
#define ARPHRD_ETHER 1            /* ethernet hardware format */
#define ARPHRD_IEEE802 6          /* token-ring hardware format */
#define ARPHRD_ARCNET 7           /* arcnet hardware format */
#define ARPHRD_FRELAY 15          /* frame relay hardware format */
#define ARPHRD_IEEE1394 24        /* firewire hardware format */
    uint16_t ar_pro;              /* format of protocol address */
    uint8_t ar_hln;               /* length of hardware address */
    uint8_t ar_pln;               /* length of protocol address */
    uint16_t ar_op;               /* one of: */
#define ARPOP_REQUEST 1           /* request to resolve address */
#define ARPOP_REPLY 2             /* response to previous request */
#define ARPOP_REVREQUEST 3        /* request protocol address given hardware */
#define ARPOP_REVREPLY 4          /* response giving protocol address */
#define ARPOP_INVREQUEST 8        /* request to identify peer */
#define ARPOP_INVREPLY 9          /* response identifying peer */
    uint8_t ar_sha[ETH_ADDR_LEN]; /* sender hardware address */
    uint8_t ar_spa[IP_ADDR_LEN];  /* sender protocol address */
    uint8_t ar_tha[ETH_ADDR_LEN]; /* target hardware address */
    uint8_t ar_tpa[IP_ADDR_LEN];  /* target protocol address */
};

/* IPv4 header */
struct ip
{
#if BYTE_ORDER == LITTLE_ENDIAN
    uint8_t ip_hl : 4, /* header length */
        ip_v : 4;      /* version */
#endif
#if BYTE_ORDER == BIG_ENDIAN
    uint8_t ip_v : 4, /* version */
        ip_hl : 4;    /* header length */
#endif
    uint8_t ip_tos;                /* type of service */
    uint16_t ip_len;               /* total length */
    uint16_t ip_id;                /* identification */
    uint16_t ip_off;               /* fragment offset field */
#define IP_RF 0x8000               /* reserved fragment flag */
#define IP_DF 0x4000               /* dont fragment flag */
#define IP_MF 0x2000               /* more fragments flag */
#define IP_OFFMASK 0x1fff          /* mask for fragmenting bits */
    uint8_t ip_ttl;                /* time to live */
    uint8_t ip_p;                  /* protocol */
    uint16_t ip_sum;               /* checksum */
    struct in_addr ip_src, ip_dst; /* source and destination address */
} __packed_ip;

/* IPv6 header */
struct ip6
{
    union
    {
        struct ip6_hdrctl
        {
            uint32_t ip6_un1_flow; /* 4-bit version, 8-bit traffic class, 20-bit flow label */
            uint16_t ip6_un1_plen; /* 16-bit payload length */
            uint8_t ip6_un1_nxt;   /* 8-bit next header */
            uint8_t ip6_un1_hlim;  /* 8-bit hop limit */
        } ip6_un1;
        uint8_t ip6_un2_vfc; /* 4-bit version, top 4-bit traffic class */
    } ip6_ctlun;
    struct in6_addr ip6_src; /* 128-bit source address */
    struct in6_addr ip6_dst; /* 128-bit destination address */
} __packed_ip6;

/*
Sample IPv6 packet showing how Wireshark decodes flow info (ip6_flow):

Frame 1: 86 bytes on wire (688 bits), 86 bytes captured (688 bits)
Ethernet II, Src: aa:aa:aa:aa:aa:aa (aa:aa:aa:aa:aa:aa), Dst: bb:bb:bb:bb:bb:bb (bb:bb:bb:bb:bb:bb)
Internet Protocol Version 6, Src: 2606:4700:4700::64, Dst: 2606:4700:4700::6400
    0110 ....                               = Version: 6
    .... 0000 0000 .... .... .... .... .... = Traffic Class: 0x00 (DSCP: CS0, ECN: Not-ECT)
    .... .... .... 0011 0101 0001 1011 0011 = Flow Label: 0x351b3
    Payload Length: 32
    Next Header: TCP (6)
    Hop Limit: 53
    Source Address: 2606:4700:4700::64
    Destination Address: 2606:4700:4700::6400
Transmission Control Protocol, Src Port: 30000, Dst Port: 60000, Seq: 1, Ack: 1, Len: 0
*/
#define ip6_flow ip6_ctlun.ip6_un1.ip6_un1_flow
#define ip6_plen ip6_ctlun.ip6_un1.ip6_un1_plen
#define ip6_nxt ip6_ctlun.ip6_un1.ip6_un1_nxt
#define ip6_hlim ip6_ctlun.ip6_un1.ip6_un1_hlim

#define IP6_FLOWLABEL_MASK (htonl(0x000fffff)) /* 20-bit flow label (network byte order) */

/* IPv4 pseudo header for computing TCP and UDP checksums */
struct ippseudo
{
    struct in_addr ippseudo_src; /* source address */
    struct in_addr ippseudo_dst; /* destination address */
    uint8_t ippseudo_pad;        /* pad, must be zero */
    uint8_t ippseudo_p;          /* protocol */
    uint16_t ippseudo_len;       /* protocol length */
};

/* IPv6 pseudo header for computing ICMPv6, TCP and UDP checksums */
struct ip6pseudo
{
    struct in6_addr ip6pseudo_src; /* 128-bit source address */
    struct in6_addr ip6pseudo_dst; /* 128-bit destination address */
    uint32_t ip6pseudo_len;        /* 32-bit upper-layer packet length */
    uint8_t ip6pseudo_zero[3];     /* 24-bit zeros */
    uint8_t ip6pseudo_nxt;         /* 8-bit next header */
};

/* ICMP header (up to checksum field only) */
struct icmphdr
{
    uint8_t icmp_type;   /* type field */
    uint8_t icmp_code;   /* code field */
    uint16_t icmp_cksum; /* checksum field */
};

/* ICMPv6 header (up to checksum field only) */
struct icmp6hdr
{
    uint8_t icmp6_type;   /* type field */
    uint8_t icmp6_code;   /* code field */
    uint16_t icmp6_cksum; /* checksum field */
};

/* TCP header */
struct tcphdr
{
    uint16_t th_sport; /* source port */
    uint16_t th_dport; /* destination port */
    uint32_t th_seq;   /* sequence number */
    uint32_t th_ack;   /* acknowledgment number */
#if BYTE_ORDER == LITTLE_ENDIAN
    uint8_t th_x2 : 4, /* (unused) */
        th_off : 4;    /* data offset in number of 32-bit words */
#endif
#if BYTE_ORDER == BIG_ENDIAN
    uint8_t th_off : 4, /* data offset in number of 32-bit words */
        th_x2 : 4;      /* (unused) */
#endif
    uint8_t th_flags;
#define TH_FIN 0x01  /* no more data from sender */
#define TH_SYN 0x02  /* synchronize sequence numbers */
#define TH_RST 0x04  /* reset the connection */
#define TH_PUSH 0x08 /* push function */
#define TH_ACK 0x10  /* acknowledgment field is significant */
#define TH_URG 0x20  /* urgent pointer field is significant */
#define TH_ECE 0x40  /* explicit congestion notification echo */
#define TH_CWR 0x80  /* congestion window reduced */
#define TH_FLAGS (TH_FIN | TH_SYN | TH_RST | TH_PUSH | TH_ACK | TH_URG | TH_ECE | TH_CWR)
    uint16_t th_win; /* window */
    uint16_t th_sum; /* checksum */
    uint16_t th_urp; /* urgent pointer */
};

/* UDP header */
struct udphdr
{
    uint16_t uh_sport; /* source port */
    uint16_t uh_dport; /* destination port */
    uint16_t uh_ulen;  /* udp length */
    uint16_t uh_sum;   /* udp checksum */
};

/*
 * Structures for bittwiste header specific options.
 */
struct eth_addr_opt
{
    uint8_t old[ETH_ADDR_LEN]; /* 48-bit Ethernet address */
    uint8_t new[ETH_ADDR_LEN];
    uint8_t flag;
};

struct ethopt
{
    struct eth_addr_opt dhost;
    struct eth_addr_opt shost;
    uint16_t eth_type;
};

struct arpopt
{
    uint16_t ar_op; /* opcode */
    uint8_t ar_op_flag;
    struct eth_addr_opt sha;         /* sender hardware address */
    uint8_t ar_old_spa[IP_ADDR_LEN]; /* sender protocol address */
    uint8_t ar_new_spa[IP_ADDR_LEN];
    uint8_t ar_spa_flag;
    struct eth_addr_opt tha;         /* target hardware address */
    uint8_t ar_old_tpa[IP_ADDR_LEN]; /* target protocol address */
    uint8_t ar_new_tpa[IP_ADDR_LEN];
    uint8_t ar_tpa_flag;
};

struct in_addr_opt
{
    struct in_addr old; /* 32-bit IPv4 address */
    struct in_addr new;
    struct in_addr netmask; /* netmask to use when randomizing address */
    uint8_t rand_bits;      /* number of bits available to the right that can be randomized */
    uint8_t flag;
};

struct ipopt
{
    uint8_t ip_ds_field; /* 6-bit DS field (first 6-bit of 8-bit type of service field) */
    uint8_t ip_ds_field_flag;
    uint8_t ip_ecn_field; /* 2-bit ECN field (last 2-bit of 8-bit type of service field) */
    uint8_t ip_ecn_field_flag;
    uint16_t ip_old_id; /* identification */
    uint16_t ip_new_id;
    uint8_t ip_id_flag;
    uint8_t ip_flag_r; /* reserved bit */
    uint8_t ip_flag_d; /* don't fragment bit */
    uint8_t ip_flag_m; /* more fragment bit */
    uint8_t ip_flags_flag;
    uint16_t ip_fo; /* fragment offset in bytes */
    uint8_t ip_fo_flag;
    uint8_t ip_old_ttl; /* time to live */
    uint8_t ip_new_ttl;
    uint8_t ip_ttl_flag;
    uint8_t ip_old_p; /* protocol */
    uint8_t ip_new_p;
    uint8_t ip_p_flag;
    struct in_addr_opt ip_src; /* options for source address */
    struct in_addr_opt ip_dst; /* options for destination address */
};

struct in6_addr_opt
{
    struct in6_addr old; /* 128-bit IPv6 address */
    struct in6_addr new;
    struct in6_addr netmask; /* netmask to use when randomizing address */
    uint8_t rand_bits;       /* number of bits available to the right that can be randomized */
    uint8_t flag;
};

struct ip6opt
{
    uint8_t ip6_ds_field; /* 6-bit DS field (first 6-bit of 8-bit traffic class field) */
    uint8_t ip6_ds_field_flag;
    uint8_t ip6_ecn_field; /* 2-bit ECN field  (last 2-bit of 8-bit traffic class field) */
    uint8_t ip6_ecn_field_flag;
    uint32_t ip6_flow_label; /* 20-bit flow label */
    uint8_t ip6_flow_label_flag;
    uint8_t ip6_old_next_header; /* 8-bit next header */
    uint8_t ip6_new_next_header;
    uint8_t ip6_next_header_flag;
    uint8_t ip6_old_hop_limit; /* 8-bit hop limit */
    uint8_t ip6_new_hop_limit;
    uint8_t ip6_hop_limit_flag;
    struct in6_addr_opt ip6_src; /* options for source address */
    struct in6_addr_opt ip6_dst; /* options for destination address */
};

struct icmpopt
{
    uint8_t icmp_type; /* type of message */
    uint8_t icmp_type_flag;
    uint8_t icmp_code; /* type sub code */
    uint8_t icmp_code_flag;
};

struct icmp6opt
{
    uint8_t icmp6_type; /* type of message */
    uint8_t icmp6_type_flag;
    uint8_t icmp6_code; /* type sub code */
    uint8_t icmp6_code_flag;
};

struct tcpopt
{
    uint16_t th_old_sport; /* source port */
    uint16_t th_new_sport;
    uint8_t th_sport_flag;
    uint16_t th_old_dport; /* destination port */
    uint16_t th_new_dport;
    uint8_t th_dport_flag;
    uint32_t th_old_seq; /* sequence number */
    uint32_t th_new_seq;
    uint8_t th_seq_flag;
    uint32_t th_old_ack; /* acknowledgment number */
    uint32_t th_new_ack;
    uint8_t th_ack_flag;
    uint8_t th_flag_c; /* CWR */
    uint8_t th_flag_e; /* ECE */
    uint8_t th_flag_u; /* URG */
    uint8_t th_flag_a; /* ACK */
    uint8_t th_flag_p; /* PSH */
    uint8_t th_flag_r; /* RST */
    uint8_t th_flag_s; /* SYN */
    uint8_t th_flag_f; /* FIN */
    uint8_t th_flags_flag;
    uint16_t th_win; /* window */
    uint8_t th_win_flag;
    uint16_t th_urp; /* urgent pointer */
    uint8_t th_urp_flag;
};

struct udpopt
{
    uint16_t uh_old_sport; /* source port */
    uint16_t uh_new_sport;
    uint8_t uh_sport_flag;
    uint16_t uh_old_dport; /* destination port */
    uint16_t uh_new_dport;
    uint8_t uh_dport_flag;
};

#endif /* !_DEF_H_ */
